### ISSUE TITLE
URL for an arbitrary day of history and navigation for next and previous days

### ISSUE THREAD

**Issue by pdurbin** [vector-im/element-web#7677]
### Description

As someone returning from vacation I would like to easily catch up on the message from the previous week or so for a room that allows anyone to read messages without logging in.

If I want to catch up on a Gitter channel, I can navigate to the URL such as https://gitter.im/gitterHQ/gitter/archives/2018/11/09 and see all the messages for that day. If I want to catch up on the Firefox IRC channel I can go to https://mozilla.logbot.info/firefox/20181109 . I don't know how to accomplish the same task for https://riot.im/app/#/room/#riot:matrix.org . Scrolling forever in a busy channel is not a satisfying user experience.

At https://matrix.to/#/!iNmaIQExDMeqdITdHH:matrix.org/$15417756832452fUMFW:cervoi.se @erdnaxeli suggested using the first message of the day as the link for that day (thanks for the suggestion!) but this could involve a lot of scrolling, which is frustrating.

@jryans confirmed for me that it's ok to open this issue (thanks!) so here I am.

### Steps to reproduce

- Go to https://riot.im/app/#/room/#riot:matrix.org
- Look for a "previous day" link or a "history by day" link.
- I can't find a way to go directly to the day I'm interested in reading.

For the Gitter and IRC examples above, here is how it looks when you pick a day to start catching up on and navigate to other days:

<img width="1202" alt="screen shot 2018-11-09 at 11 14 36 am" src="https://user-images.githubusercontent.com/21006/48274491-7139f200-e411-11e8-8643-08d30711ccc6.png">

<img width="1125" alt="screen shot 2018-11-09 at 11 13 52 am" src="https://user-images.githubusercontent.com/21006/48274492-71d28880-e411-11e8-82da-468f75e6c490.png">

### Version information

- **Platform**: web

For the web app:

- **Browser**: All.
- **OS**: All.
- **URL**: Any Riot room but https://riot.im/app/#/room/#riot:matrix.org for example

**Comment by lampholder** [comment]
There's a chance this could be implemented with the search API (if it supports date range search terms), otherwise we might need to make a Synapse change. This does sound like a nice feature, though.

**Comment by VasoVV** [comment]
Today I've trying to find the massage I'd sent at March. I remember the date, I remember content… but it is encrypted room… no search works… no jump to date works… so I'm scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling and scrolling…

My mouse hates Riot. My mouse suffers. Let's protect animals! Let's resolve this issue!!

So I think, if search in e2e rooms doesn't work the "jump to day" function will be good enough workaround. I wonder why this issue is not popular and here are only two participants.

**Comment by surli** [comment]
Anything new on this topic? I just have the exact same issue as the first post author: returning from 2 weeks of vacations, trying to get back on some specific discussions related to some emails dates, and no way to browse back on a day based on the date. 

Moreover, when I scroll back I can check on the horizontal bar that separate days, but I only have the name on the day for the current week so not so easy to verify which date we were last saturday and check the associated date on mails...

**Comment by pdurbin** [comment]
You can go to https://view.matrix.org/alias/%23riot:matrix.org for example) and click "Load older messages" over and over. Here's a screenshot:

<img width="1577" alt="Screen Shot 2019-10-03 at 7 44 02 AM" src="https://user-images.githubusercontent.com/21006/66124136-9d260f00-e5b1-11e9-8d40-1d9d77dee76d.png">

I would still rather have "next day" and "previous day" buttons but this is better than nothing. Here's the project that powers the pages above: https://github.com/matrix-org/matrix-static

**Comment by surli** [comment]
Thanks for the hint, but we use our own instance of matrix, so we'd need to install our own instance of matrix-static too AFAIU.

I still think it would be really useful to have a proper way to browse messages by date on riot itself, not only previous/next days, but with a full calendar.

**Comment by rastapopougros** [comment]
Same use case as @VasoVV for me, I kept scrolling for many *minutes* (it's not just an image, I really scrolled entire minutes) before finding the history I wanted. And it was just 1 month ago… imagine if I wanted to find what we said last year… :(

**Comment by Biep** [comment]
This would also need absolute times/dates.  "The first message of the day" is client-dependent.
So it would go nicely with #12285.

**Comment by MurzNN** [comment]
Matrix client must know client timezone and can convert time to UTC for server query. Also client can provide choosing other timezone. So at first we need to implement server-side support in SPEC for jump to specific timestamp.

**Comment by buzmeg** [comment]
Giving this a bump as scrolling back through history is infuriating to me as well.

> Matrix client must know client timezone and can convert time to UTC for server query.

This is putting the cart before the horse.  Most people would be willing to type in the UTC directly as long as it scrolled back properly.

This is exacerbated by the fact that everything in the universe causes a "scroll to bottom" if there is even a slight hiccup.   It's quite maddening.

**Comment by rgpublic** [comment]
I would find it very cool if there was a small icon next to the day separator ("Thu, July 9 2020") in the chat history that would open a small calendar where you can select a date and then it jumps/scrolls right to that position. It used to be quite terrible already, but in recent times (perhaps with the new Element?) the scrolling has gotten even slower - it's a nightmare now to look for a specific date in the chat history when you have to scroll through hundreds of pages.

**Comment by buzmeg** [comment]
Any progress?

**Comment by richvdh** [comment]
I don't think there's actually any way to use this (yet), at least on mainline Element and Synapse? So it should probably not be closed?

**Comment by t3chguy** [comment]
@richvdh there's a labs flag for jump to date functionality if your HS exposes support for it.

**Comment by richvdh** [comment]
I'm generally unconvinced that we should consider a task "done" if it requires enabling experimental options (either client-side or server-side), but we can have that discussion another day.

**Comment by rgpublic** [comment]
Phew. That was tricky. In case anyone else wants to try this out. In /etc/matrix-synapse/homeserver.yaml on your server you need to add these lines at the end:

```
experimental_features:
   msc3030_enabled: true
```
Please note: There is no dash in front of "msc3030_enabled: true" like you'll sometimes find in examples on the web.
Don't forget to restart the service: `systemctl restart matrix-synapse.service`

Next, on the client you need to change "showLabsSettings" from false to true in /etc/element/config.json. What confused me a lot is that even when this setting is still false you'll nevertheless see 2 Lab experiments under Settings-> Labs (Video rooms and Threads). These are always there no matter what the setting is. If you sucessfully enabled the flag (don't forget to restart the app, make sure it is really closed and not running in the tray) you will see a lot of on/off switches below these two experiments when you scroll down. Now, there you can enable the "Jump to date" flag. After you have enabled this flag, you'll notice a drop down arrow next to the date divider in the timeline. 

Hope this helps anyone - took me a while to figure this out.

**Comment by rastapopougros** [comment]
Thanks for the tips. But.

I'm agree that a very (very) hidden feature, is not the same as an added supported feature. So it's not "finished" to be integrated, just in a waiting room. So not really closed.

Also important, not anyone has it's own server to play with : the request for this feature must be resolved for just simple users who need to search in history, not scrolling dozens of minutes to find a thing 2 years ago.

**Comment by t3chguy** [comment]
I suggest petitioning [Synapse](https://github.com/matrix-org/synapse/) to stabilise their support of the feature so that clients can have a chance to enable it more widely.

**Comment by MadLittleMods** [comment]
For those wanting to add umpf to stabilizing support, give [MSC3030](https://github.com/matrix-org/matrix-spec-proposals/pull/3030) a 👍 (and feedback if you have any). Once the MSC lands, we can make the API stable in Synapse and go generally available with all of the jump to date features in Element.

Jump to date in Element is totally usable today with the only thing holding it back are the experimental endpoints contingent on the MSC landing.

For those wanting to use it today, @rgpublic laid out some [great steps to enable support](https://github.com/vector-im/element-web/issues/7677#issuecomment-1284042644) in Synapse and Element. If you have a `matrix.org` account, then this won't be available because it doesn't have MSC3030 enabled.

**Comment by pdurbin** [comment]
What's gotten a bit lost from the original request is the part about a URL.

Specifically, I was hoping for a URL to navigate to for a particular date. The examples I gave were:

- https://gitter.im/gitterHQ/gitter/archives/2018/11/09
- https://mozilla.logbot.info/firefox/20181109

(Neither of these examples work now, which is a little sad. 😞 )

While a `/jumptodate` commands sounds useful, I'm actually much more excited about URLs with dates in them that @MadLittleMods recently demo'ed at https://youtu.be/UT6KSEqDUf8?t=581

Here's a screenshot:

<img width="965" alt="Screen Shot 2022-10-22 at 5 50 52 PM" src="https://user-images.githubusercontent.com/21006/197363655-9122b343-2d4e-419a-ae07-f850ac0a78ce.png">

What I see is a URL that looks something like this:

http://localhost:3050/!inscrutableStringHere:my.synapse.server/date/2022/02/24

There's even a nice calendar picker, just like Gitter has. (Or had, perhaps, since as mentioned, the Gitter link with a date no longer works.)

The feature is apparently called "Matrix public archive" and here's the writeup (also by @MadLittleMods): https://matrix.org/blog/2022/08/05/this-week-in-matrix-2022-08-05#matrix-public-archive-website

You can follow the progress at https://github.com/matrix-org/matrix-public-archive . Thanks, @MadLittleMods!

**Comment by MadLittleMods** [comment]
> (Neither of these examples work now, which is a little sad. 😞 )

For the Gitter one, the room URI was renamed but still works 🙂: https://gitter.im/gitter/gitter/archives/2018/11/09 (too bad the rename redirect doesn't work for archives)


> What's gotten a bit lost from the original request is the part about a URL.

Feel free to re-open this issue or create another issue specifically focusing on this.

This was obviously part of the issue title and referenced all over the issue but the reproduction steps mention "I can't find a way to go directly to the day I'm interested in reading." which more aligned with jump to date functionality.

Thanks for linking everything related to [Matrix Public Archive](https://github.com/matrix-org/matrix-public-archive) which will solve this use case at least outside of Element 🏋️

**Comment by pdurbin** [comment]
@MadLittleMods thanks, no need to reopen this issue. I'm tracking my use case [elsewhere](https://github.com/IQSS/chat.dataverse.org/issues/15). I can't wait until we can start using Matrix Public Archive! 🎉

**Comment by MadLittleMods** [comment]
> Feel free to re-open this issue or create another issue specifically focusing on [the original request about a URL to navigate by day.]

Created https://github.com/vector-im/element-web/issues/24914 to track this ⏩
